c2rust-analyze/src/borrowck/type_check.rs

```diff
@@ -33,10 +37,92 @@ impl<'tcx> TypeChecker<'tcx, '_> {
     }
 
     pub fn visit_place(&mut self, pl: Place<'tcx>) -> LTy<'tcx> {
-        let mut lty = self.local_ltys[pl.local.index()];
+        let mut lty: LTy = self.local_ltys[pl.local.index()];
+        let map: &mut HashMap<OriginKind, Origin> = &mut lty
+            .label
+            .origin_params
+            .map(|params| HashMap::from_iter(params.to_vec()))
+            .unwrap_or_default();
+        let mut adt_func = |base_adt_def: AdtDef, field: Field, field_ty: Ty<'tcx>| {
```

Comment:
This strategy of keeping a persistent map that gets updated over time seems potentially fragile, particularly since we eventually need to support recursive or mutually-recursive types.  But I think the actual operation being performed here is essentially just a standard type substitution, e.g. replace `'b` with `'0` in the declared field type `Foo<'b>`, where the mapping `'b -> '0` is obtained by matching up the formal parameters of `struct Bar<'b>` with the actual arguments of the base type `Bar<'0>`.  I think if you modify `util::lty_project` to pass through the base `LTy` into `adt_func` (since the base `LTy` is the one that contains the concrete origin `'0`), then you can do the whole substitution inside `adt_func` without carrying any mutable state.  (You already have `self.adt_metadata`, from which you can obtain the formal parameters of the ADT and the declared field type.)  The one quirk here is that the "declared field type" will have to be labeled with `OriginKind`s rather than `Label`s/`Origin`s, since it needs to refer to the origin parameters rather than concrete origins.

Reply:
addressed in edff552b